go/plugins/internal/gemini/cache.go

```diff
@@ -0,0 +1,224 @@
+// genai.CachedContent that should be cached.
+// This is where you decide what goes into the cache: large documents, system instructions, etc.
+func prepareCacheContent(
+	request *ai.ModelRequest,
+	model string,
+) (*genai.CreateCachedContentConfig, error) {
+	cacheEndIdx, ttl, err := findCacheMarker(request)
+	if err != nil {
+		return nil, err
+	}
+	// no cache mark found
+	if cacheEndIdx == -1 {
+		return nil, err
+	}
+	// index out of bounds
+	if cacheEndIdx < 0 || cacheEndIdx >= len(request.Messages) {
+		return nil, fmt.Errorf("end of cached contents, index %d is invalid", cacheEndIdx)
+	}
+
+	var messagesToCache []*genai.Content
+	for i := cacheEndIdx; i >= 0; i-- {
+		m := request.Messages[i]
+		if m.Role == ai.RoleSystem {
+			continue
+		}
+		parts, err := convertParts(m.Content)
+		if err != nil {
+			return nil, err
+		}
+		messagesToCache = append(messagesToCache, &genai.Content{
+			Parts: parts,
+			Role:  string(m.Role),
+		})
+	}
+
+	key := calculateCacheKey(messagesToCache)
+
+	fmt.Printf("cache key: %s\n\n", key)
+	return &genai.CreateCachedContentConfig{
+		DisplayName: key,
+		TTL:         strconv.Itoa(ttl) + "s",
+		Contents:    messagesToCache,
+	}, nil
+}
+
+// validateContextCacheRequest checks for supported models and checks if Tools
+// are being provided in the request
+func validateContextCacheRequest(request *ai.ModelRequest, modelVersion string) error {
+	if modelVersion == "" || !slices.Contains(ContextCacheSupportedModels[:], modelVersion) {
+		return fmt.Errorf("%s", INVALID_ARGUMENT_MESSAGES.modelVersion)
+	}
+	if len(request.Tools) > 0 {
+		return fmt.Errorf("%s", INVALID_ARGUMENT_MESSAGES.tools)
+	}
+	for _, m := range request.Messages {
+		if m.Role == ai.RoleSystem {
+			return fmt.Errorf("%s", INVALID_ARGUMENT_MESSAGES.systemPrompt)
+		}
+	}
+
+	return nil
+}
+
+// findCacheMarker finds the cache mark in the list of request messages.
+// All of the messages preceding this mark will be cached
+func findCacheMarker(request *ai.ModelRequest) (int, int, error) {
+	for i := len(request.Messages) - 1; i >= 0; i-- {
+		m := request.Messages[i]
+		if m.Metadata == nil {
+			continue
+		}
+
+		cacheVal, exists := m.Metadata["cache"]
+		if !exists || cacheVal == nil {
+			continue
+		}
+
+		c, ok := cacheVal.(map[string]any)
+		if !ok {
+			return -1, 0, fmt.Errorf("cache metadata should be map but got: %T", cacheVal)
+		}
+
+		if t, ok := c["ttlSeconds"].(int); ok {
+			return i, t, nil
+		}
+
+		return -1, 0, fmt.Errorf("invalid type for cache ttlSeconds, expected int, got %T", c["ttlSeconds"])
+	}
+	return -1, 0, nil
+}
+
+// lookupCache retrieves a *genai.CachedContent from a given cache key
+func lookupCache(ctx context.Context, client *genai.Client, key string) (*genai.CachedContent, error) {
+	if key == "" {
+		return nil, fmt.Errorf("empty cache key detected")
+	}
+
+	var nextPageToken string
+	for {
+		page, err := client.Caches.List(ctx, &genai.ListCachedContentsConfig{
```

Comment:
Why not `client.Caches.Get(...)`?

Reply:
This is not a practical solution. You can have many requests and many caches active. This lookup is O(n), it will be super slow for any prod uses with traffic. We need a mechanism to return `Name` back to user control so that it can be passed through to subsequent requests, both in Go and in JS.

In any case, `Caches.All()` returns an iterator over all caches available which is better to use here if we're going to brute force it. You should file an issue if you're going to check this in.

Reply:
As per the suggestion in discord, we are now using `WithCacheName()` helper to fetch in constant time the required cache